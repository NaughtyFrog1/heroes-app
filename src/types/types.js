const types = {
  login: '[auth] login',
  logout: '[auth] logout',
}

export default types

/**
 * Este objeto va a contener todos los types de la aplicaci√≥n.
 * Entre corchetes indicamos a que grupo de acciones pertenece el type
 */