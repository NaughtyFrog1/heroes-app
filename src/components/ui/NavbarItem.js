import React from 'react'
import PropTypes from 'prop-types'
import { NavLink } from 'react-router-dom'

const NavbarItem = ({ to, children }) => {
  return (
    <li className="nav-item">
      <NavLink activeClassName="active" className="nav-link" exact to={to}>
        {children}
      </NavLink>
    </li>
  )
}

NavbarItem.propTypes = {
  to: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

export default NavbarItem
