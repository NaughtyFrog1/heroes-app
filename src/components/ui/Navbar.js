import React, { useContext } from 'react'
import { Link, useHistory } from 'react-router-dom'
import AuthContext from '../../auth/AuthContext'
import types from '../../types/types'
import NavbarItem from './NavbarItem'

export const Navbar = () => {
  const { user, dispatch } = useContext(AuthContext)
  const history = useHistory()

  const handleClickLogout = () => {
    dispatch({ type: types.logout })
    history.replace('login')
  }

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <div className="container-fluid">
        <Link className="navbar-brand" to="/">
          HeroesApp
        </Link>
        <button
          className="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarNavDropdown"
          aria-controls="navbarNavDropdown"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span className="navbar-toggler-icon"></span>
        </button>

        <div className="collapse navbar-collapse" id="navbarNavDropdown">
          <ul className="navbar-nav">
            <NavbarItem to="/marvel">Marvel</NavbarItem>
            <NavbarItem to="/dc">DC</NavbarItem>
            <NavbarItem to="/search">Search</NavbarItem>
          </ul>
          <ul className="navbar-nav ms-auto">
            <span className="nav-item nav-link text-muted fw-light">
              {user.name}
            </span>
            <button className="nav-item nav-link btn" onClick={handleClickLogout}>
              Logout
            </button>
          </ul>
        </div>
      </div>
    </nav>
  )
}
