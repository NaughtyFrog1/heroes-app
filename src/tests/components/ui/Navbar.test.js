import { mount } from 'enzyme'
import { MemoryRouter, Router } from 'react-router'
import AuthContext from '../../../auth/AuthContext'
import { Navbar } from '../../../components/ui/Navbar'
import types from '../../../types/types'

describe('tests in Navbar', () => {
  // Creamos un history mock para poder testearlo, ya que es lo que usa el useHistory
  // historyMock estÃ¡ formado por todos los elementos que necesita el useHistory para funcionar
  const historyMock = {
    location: {},
    push: jest.fn(),
    listen: jest.fn(),
    replace: jest.fn(),
    createHref: jest.fn(),
  }

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: true,
      name: 'John Doe',
    },
  }

  afterEach(() => {
    // Al finalizar cada una de las pruebas es conveniente limpiar los mocks
    jest.clearAllMocks()
  })

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter>
        {/* Pasamos el historyMock al router */}
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </MemoryRouter>
    </AuthContext.Provider>
  )

  test('should render properly', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.navbar__name').text().trim()).toBe('John Doe')
  })

  test('should call logout and use history', () => {
    wrapper.find('.navbar__logout').prop('onClick')()
    expect(contextValue.dispatch).toHaveBeenCalledWith({ type: types.logout })
    expect(historyMock.replace).toHaveBeenCalledWith('login')
  })
})
