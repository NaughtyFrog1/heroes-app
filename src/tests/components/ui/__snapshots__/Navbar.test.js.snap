// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in Navbar should render properly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      onClick={[Function]}
    >
      HeroesApp
    </a>
    <button
      aria-controls="navbarNavDropdown"
      aria-expanded="false"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      data-bs-target="#navbarNavDropdown"
      data-bs-toggle="collapse"
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      id="navbarNavDropdown"
    >
      <ul
        className="navbar-nav"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Marvel
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            DC
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Search
          </a>
        </li>
      </ul>
      <ul
        className="navbar-nav ms-auto"
      >
        <span
          className="nav-item nav-link text-muted fw-light navbar__name"
        >
          John Doe
        </span>
        <button
          className="nav-item nav-link btn navbar__logout"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </div>
</nav>
`;
